exporter/collector/metrics.go

```diff
@@ -271,44 +365,254 @@ func (me *MetricsExporter) PushMetrics(ctx context.Context, m pmetric.Metrics) e
-				if summary, ok := detail.(*monitoringpb.CreateTimeSeriesSummary); ok {
-					failedPoints = int(summary.TotalPointCount - summary.SuccessPointCount)
-					succeededPoints = int(summary.SuccessPointCount)
-				}
+			if !(s.Code() == codes.DeadlineExceeded || s.Code() == codes.Unavailable) {
+				break
 			}
+			me.obs.log.Error("retryable error, retrying request")
+			backoff = 1 << i
+			if time.Duration(backoff)*time.Second >= me.wal.maxBackoff {
+				break
+			}
+			time.Sleep(time.Duration(backoff) * time.Second)
+		}
 
-			// always record the number of successful points
-			recordPointCountDataPoint(ctx, succeededPoints, "OK")
-			if failedPoints > 0 {
-				recordPointCountDataPoint(ctx, failedPoints, st)
+		lastIndex, indexErr := me.wal.LastIndex()
+		if indexErr != nil {
+			return indexErr
+		}
+		if readIndex == lastIndex && len(req.String()) > 0 {
+			// This indicates that we are trying to truncate the last item in the WAL.
+			// If that is the case, write an empty request so we can truncate the last real request
+			// (the WAL library requires at least 1 entry).
+			// Doing so prevents double-exporting in the event of a collector restart.
+			emptyReq := &monitoringpb.CreateTimeSeriesRequest{}
+			bytes, bytesErr := proto.Marshal(emptyReq)
+			if bytesErr != nil {
+				return bytesErr
 			}
+
+			err = me.wal.Write(writeIndex+1, bytes)
 			if err != nil {
-				errs = append(errs, fmt.Errorf("failed to export time series to GCM: %+v", s))
+				return err
 			}
 		}
+		// move read index forward if non retryable error (or exported successfully)
+		err = me.wal.TruncateFront(readIndex + 1)
+		if err != nil {
+			return err
+		}
+		return nil
 	}
-	if len(errs) > 0 {
-		return multierr.Combine(errs...)
+	return err
+}
+
+// watchWAL watches the WAL directory for a write then returns to the
+// continuallyPopWAL() loop.
+func (me *MetricsExporter) watchWALFile(ctx context.Context) error {
+	me.goroutines.Add(1)
+	defer me.goroutines.Done()
+	walWatcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+	err = walWatcher.Add(me.wal.path)
+
+	if err != nil {
+		return err
+	}
+	watchCh := make(chan error)
+	var wErr error
+	go func() {
+		defer func() {
+			watchCh <- wErr
+			close(watchCh)
+			walWatcher.Close()
+		}()
+
+		select {
+		case <-me.shutdownC:
+			return
+		case <-ctx.Done():
+			wErr = ctx.Err()
+			return
+
+		case event, ok := <-walWatcher.Events:
+			if !ok {
+				return
+			}
+			switch event.Op {
+			case fsnotify.Remove:
+				wErr = fmt.Errorf("WAL file deleted")
+			case fsnotify.Rename:
+				wErr = fmt.Errorf("WAL file renamed")
+			case fsnotify.Write:
+				wErr = nil
+			}
+
+		case watchErr, ok := <-walWatcher.Errors:
+			if ok {
+				wErr = watchErr
+			}
+		}
+	}()
+	err = <-watchCh
+	return err
+}
+
+func (me *MetricsExporter) runWALReadAndExportLoop(ctx context.Context) {
+	defer me.goroutines.Done()
+	runCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
```

Comment:
We're not using this cancel anywhere, I think this code is a noop

Reply:
sounds good, removed this and changed the `runCtx` to just `ctx`